### PR TITLE
Move yetus into dockerfile with make target

### LINKED ISSUE #183: Yetus should be accessible outside pull requests
Please implement a way to run Yetus checks on our code changes before we create a pull request. Right now, we submit a PR, and then come to know about Yetus errors as part of the checks.  We end up wasting circleCI bandwidth in iterating over Yetus failures, and becomes a time hurdle in code reviews.

What would be good is to have a way to run Yetus on our local EVE repository, so that we can validate Yetus in our laptop before we even create a PR.

### PR DESCRIPTION
**Pull request by deitch**
Signed-off-by: Avi Deitcher <avi@deitcher.net>

Fixes #183

### CODE REVIEW

**Comment by deitch** [comment]
Running it locally doesn't actually work. It reports that it is missing something in the command-line args. @rvs do you know why?

**Comment by deitch** [comment]
I will admit, I do not know enough about yetus yet ... (pun definitely intentional)

**Comment by deitch** [comment]
Same problem when running in CI:

```
Running yetus, results saved to /tmp/yetus-out
docker run --rm -v /root/project:/src -v /tmp/yetus-out:/tmp/yetus-out lfedge/eve-yetus:build
WARNING: Found both spotbugs and findbugs. Disabling findbugs.
test-patch [OPTIONS] patch

Where:
  patch is a file, URL, or bugsystem-compatible location of the patch file
```

I obviously missed something converting from how it was running before, but not sure what?

File: .circleci/config.yml [hunk 1]
```diff
@@ -66,25 +55,9 @@ jobs:
           command: echo "bootstrap" > /tmp/yetus-out/bootstrap
       - store_artifacts:
           path: /tmp/yetus-out
-      - run:
+      - run:
```
Comment (rvs) [inline .circleci/config.yml]: spurious whitespace change

File: Makefile [hunk 2]
```diff
@@ -120,6 +122,16 @@ clean:
 	rm -rf $(DIST) pkg/pillar/Dockerfile pkg/qrexec-lib/Dockerfile pkg/qrexec-dom0/Dockerfile \
 	       images/installer.yml images/rootfs.yml
 
+yetus-image:
+	docker build -t $(YETUS_IMG) ./yetus
+
+yetus: yetus-image
```
Comment (rvs) [inline Makefile]: Please make sure to add yetus target to the help section of the Makefile.

And speaking of help section -- I'm not clear at all about what this target is expected to do. The way we use yetus today is for *patch* (pull request) testing. It is not clear to me at all -- what is the notion of a *patch* when you call this target and what base it has to be applied to.

**Comment by aw-was-here** [comment]
This isn't working because there is no input as to what to actually process being given to test-patch:

```
test-patch [OPTIONS] patch

Where:
  patch is a file, URL, or bugsystem-compatible location of the patch file
```

i.e., test-patch is asking where the patch it is supposed to analyze is.

There are some pre-defined environment variables that the Circle CI environment provides.  Apache Yetus uses those variables to determine which branch and/or PR to process. From those variables it then calls the Github API to pull down the patch, does some analysis, applies the patch, etc.  With this change, none of those environment variables are injected into the environment and so test-patch is correctly confused as what it is supposed to do and asking for help.

If the intent is to run this locally, then the same problem applies, but the input is different.  Instead of environment variables, there are two potential choices:

1.  it should be given the location of a 'git format-patch' file and a pristine version of the source tree (because it's going to try and modify this one)
2. it should run either as qbt or as test-patch --empty-patch to do a full source tree analysis as it does for a branch build.  (The semantics of which one to use depends upon if there is going to ever be patch information provided on the command line.  qbt will prefer full builds, test-patch will honor the patch file/URL/whatever)

With more information, I can give more specific advice.  It's probably also worth mentioning the http://yetus.apache.org/downloads/ page, which talks about local homebrew and yetus-dl.  The way most people run local invocations is to use one of these methods + a wrapper script with whatever flags the project needs.  e.g., https://github.com/apache/hadoop/blob/trunk/dev-support/bin/test-patch  (although Hadoop is a slightly bad example because they have a personality file that defines a lot of their custom bits).

File: yetus/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,29 @@
+FROM apache/yetus:master
```
[the remark this answers is not in the document]
Comment (deitch) [reply yetus/Dockerfile]: Done.

File: .circleci/config.yml (hunk 1 above)
Comment (deitch) [reply .circleci/config.yml]: Done, thanks

File: Makefile (hunk 2 above)
Comment (deitch) [reply Makefile]: > Please make sure to add yetus target to the help section of the Makefile.

Done

> what is the notion of a patch when you call this target and what base it has to be applied to

Ah, now that is a good question. 

The goal here is to be able to run it locally (secondary benefit is simplifying circleci config so it is a simple `run`, with everything else portable, but very much secondary).

I hadn't considered that it applies to patches themselves. I would have expected it to be like a lint or similar, which just applies to _everything_, not just changes. I never actually realized that it is a diff checker. I am guessing we can apply it to the diff from the base of the current branch, but let's resume this conversation below with AW.

**Comment by deitch** [comment]
Thanks @aw-was-here this is helpful. Per the above conversation, I was under the impression that this just ran on the local tree, validating format, lint, etc. for the _entire_ source tree, not just the patch. When I saw the error, it dawned upon me that this is how it works.

What we are trying to do is fulfill a request from some of the developers to run yetus locally, so that they can see results without running a PR. It shortens the feedback loop for them. Not everyone likes it that way (@rvs prefers fire-and-forget, let CI complain, fix it, repush; I prefer local), but if devs say, "it makes my life easier and faster", then trying to accommodate.

Let's take this in two distinct steps:

1. What env vars do I need to pass into the `docker run` from the CI environment to get it to work as it does today, just as a `docker run` rather than local?
1. How should I run it locally when _not_ in CI?

For the second one, I think we have two options:

* run on the entire source tree
* run on the diff between the root of the current branch and `HEAD`, which is roughly equivalent to a PR.

Separately, does it have to use the GitHub API to download? Can it use a local tree? I don't think devs will want to invoke the GitHub API when running locally.

Thanks for helping here.

**Comment by rvs** [comment]
Hey @deitch -- I think your biggest misconception is how ready our source tree is for the non-patch run of yetus. Basically -- we've got so much crap that yetus is going to complain about (most of it rightfully so, btw) that our only option is running in on patches only.

That puts a kibosh on your option #1 

So you're left with #2 -- which is simply an equivalent of git format-patch origin/master.XXX and later giving the output as a filename to yetus

File: Makefile [hunk 4]
```diff
@@ -120,6 +122,16 @@ clean:
...
 	       images/installer.yml images/rootfs.yml
 
+yetus-image:
+	docker build -t $(YETUS_IMG) ./build-tools/src/yetus
+
+yetus: yetus-image
+	@echo Running yetus, results saved to $(YETUS_OUT)
+	@mkdir -p $(YETUS_OUT)
+	@echo "bootstrap" > $(YETUS_OUT)/bootstrap
+	docker run --rm -v $(PWD):/src -v $(YETUS_OUT):/tmp/yetus-out $(YETUS_IMG)
```
Comment (rvs) [inline Makefile]: After chatting with @aw-was-here yesterday I wanted to suggest that you may have better luck with CircleCI if you simply pass all of the ENV variables down to docker execution. E.g.:

docker run --rm -v $(PWD):/src -v $(YETUS_OUT):/tmp/yetus-out --env-file <(env) $(YETUS_IMG)

Or something along these lines @deitch

Comment (deitch) [reply Makefile]: We can do that, it feels "messy" though. Well, anything to get us started. Make it work, then make it better.

Comment (deitch) [reply Makefile]: Try this.

**Comment by aw-was-here** [comment]
Take a look at https://github.com/aw-was-here/eve/tree/aw-standalone-test-patch .  I haven't tried this under CircleCI but it might work there too. :)

**Comment by deitch** [comment]
I tried passing the entire env in, it actually does run, but gets the error:

```
docker run --rm -v /root/project:/src -v /tmp/yetus-out:/tmp/yetus-out --env-file /tmp/yetus-env lfedge/eve-yetus:build
ERROR: Excluded paths file (.yetus-excludes}) does not exist!
```

So pretty close, but not sure what the issue is. @aw-was-here , is there any way to tell the bash scripts to run `set -x`, any env var or similar?

> What this does is download yetus outside of the container. From there, it launches test-patch in "default to full build mode" using a modified version of @deitch's container definition

Does that mean that we need its dependencies and scripts in the local environment? And would it differ between running it on Mac vs linux vs Windows vs different distros vs etc? Definitely want to isolate all of those in a container.

Heh, I am noticing this is turning out to be harder than we expected. :-)

**Comment by aw-was-here** [comment]
> Does that mean that we need its dependencies and scripts in the local environment?

The wrapper script will download just yetus, run everything in the Dockerfile that has all of the dependencies,take care of all of the volume mounts, etc.  The Makefile just needs to call that script and not do anything else. In other words, this should make it much much easier to get to your "want to run Yetus locally" goal.

> And would it differ between running it on Mac vs linux vs Windows vs different distros vs etc?

I tested on Mac. Windows, of course, is always problematic but this should work in WSL.  Multiple architectures also gets weird, but that's mainly a docker problem.

**Comment by deitch** [comment]
> The wrapper script will download just yetus, run everything in the Dockerfile that has all of the dependencies,take care of all of the volume mounts, etc. The Makefile just needs to call that script and not do anything else. In other words, this should make it much much easier to get to your "want to run Yetus locally" goal

Got it.

If I am going to download a script, then why wouldn't I just put that complex script as the entrypoint in the Dockerfile itself, and just run `docker run -v .... image`? Or is that what it is doing?

**Comment by rvs** [comment]
Hey @deitch now that @aw-was-here PR landed, should this be closed or reworked?

**Comment by deitch** [comment]
I am going to close this. It serves the same purpose. If the team that wanted to run yetus locally still has issues and prefers to have the entire thing run in a container, rather than the make->script->docker that #195 gives us, we can cherry-pick pieces out of this if needed. 

I wouldn't spend any more time on this now that #195 is in, unless someone explicitly tells us, "what we have isn't good enough and here is why." :-)

**Pull request closed without merging**